src/bin/psql/adhoc.c

```diff
@@ -0,0 +1,135 @@
+#include "postgres_fe.h"
+#include "adhoc_compat.h"
+#include "rowmap.h"
+#include "rowstream.h"
+#include "screen.h"
+#include "model.h"
+
+#include <curses.h>
+#include <utils/rbtree.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <poll.h>
+#include <signal.h>
+
+/*
+ * pipeline adhoc query client
+ */
+
+typedef struct App
+{
+	Model* model;
+	Screen* screen;
+} App;
+
+volatile int keep_running = 1;
+void sighandle(int);
+
+void sighandle(int x)
+{
+	keep_running = 0;
+}
+
+void row_callback(void* ctx, int type, Row* row);
```

Comment:
This is done on purpose - the callback is defined generically in RowStream.

Another way to do this would be to cast the function ptr, but I think this reads cleaner.